### PR TITLE
Store credentials per instance to allow multiple clients

### PR DESCRIPTION
**Pull request by iliran11**
solves #6 

* `base/base.ts` has changed its name and is now `/base/transport.ts` to better reflect this file is about transporting data to twitter API.
* The methods of `base/base.ts` are exactly the same in the new file - I was just moving around the functions inside the class.
*  Credentials were store as private constants of a file `/base/base.ts`.  Now the new module `/base/transport.ts` export a class which can be instantiated.
* the credentials are encapsulated inside each Transport instance
* the main class `TwitterClient` creates a new instance called `transport` and passes it to all of the clients.
* Each client now requires Transport instance and calling class methods of this instance.

thanks,
[@LaptopCouch](https://twitter.com/LaptopCouch)

### CODE REVIEW

File: src/base/transport.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import OAuth from 'oauth';
+import IClientOptions from './IClientOptions';
+import Cache from './Cache';
+import { formatURL } from './utils';
+
+class Transport {
+  private oauth: OAuth.OAuth;
+
+  private cache?: Cache;
+  private credentials: IClientOptions & { [key: string]: any };
```
Comment (Silind) [inline src/base/transport.ts:10]: Let's have another empty line here before `constructor` :slightly_smiling_face:

File: src/base/transport.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    this.oauth = new OAuth.OAuth(
+      'https://api.twitter.com/oauth/request_token',
+      'https://api.twitter.com/oauth/access_token',
+      this.credentials.apiKey,
+      this.credentials.apiSecret,
+      '1.0A',
+      null,
+      'HMAC-SHA1'
+    );
+    if (!options?.disableCache) {
```
Comment (Silind) [inline src/base/transport.ts:22]: I would suggest an empty line here as well, before the if statement :slightly_smiling_face:

File: src/base/transport.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      this.cache = new Cache(options?.ttl, options.maxByteSize);
+    }
+  }
+  updateOptions(options: Partial<IClientOptions>) {
```
Comment (Silind) [inline src/base/transport.ts:26]: Same here - let's have an extra empty line before each new method :blush: 

Also, to stick with the convention of the rest of the codebase, let's keep a `public` access modifier.
```suggestion
  public updateOptions(options: Partial<IClientOptions>) {
```

File: src/base/transport.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const { apiKey, apiSecret, ...rest } = options;
+    const cleanOptions = rest as { [key: string]: any };
+
+    Object.keys(cleanOptions).forEach((key: string) => {
+      if (cleanOptions[key]) {
+        this.credentials[key] = cleanOptions[key];
+      }
+    });
+  }
+  async doGetRequest<T>(url: string): Promise<T> {
```
Comment (Silind) [inline src/base/transport.ts:36]: Empty line + public :slightly_smiling_face: 

```suggestion
  public async doGetRequest<T>(url: string): Promise<T> {
```

File: src/base/transport.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+          const result = JSON.parse(body.toString());
+
+          this.cache?.add(url, result);
+          resolve(result);
+        }
+      );
+    });
+  }
+  async doPostRequest<T>(url: string, body?: any): Promise<T> {
```
Comment (Silind) [inline src/base/transport.ts:84]: Empty line + public :slightly_smiling_face: 
```suggestion
  public async doPostRequest<T>(url: string, body?: any): Promise<T> {
```

File: src/generator/writeClients.ts [hunk 6]
```diff
@@ -89,8 +89,16 @@ function writeClients(dictionary: IReferenceDirectory[]) {
     interfacesTypesImports.forEach((type) => {
       clientFile += `import ${type} from '../interfaces/types/${type}Types';\n`;
     });
-
+    clientFile += `import Transport from '../base/transport'\n`
     clientFile += `\nclass ${fileName}Client {\n`;
+    clientFile += `   
+    private transport: Transport;
+    constructor(transport: Transport) {
```
Comment (Silind) [inline src/generator/writeClients.ts:96]: Let's add an empty line here as well, between the `transport` member and the `constructor` :smile:

File: src/generator/writeClients.ts [hunk 7]
```diff
@@ -89,8 +89,16 @@ function writeClients(dictionary: IReferenceDirectory[]) {
...
+      if (!transport) {
+        throw Error('Transport class needs to be provided.');
+      }
+      this.transport = transport;
```
Comment (Silind) [inline src/generator/writeClients.ts:100]: And a final empty line here before assigning `this.transport` :smile:

**Review by Silind — CHANGES_REQUESTED** [review]
You are on a row, man! :fire: 
It's highly appreciated!!

I only have a couple of small things regarding some extra empty lines :blush:

File: src/base/transport.ts (hunk 1 above)
Comment (iliran11) [reply src/base/transport.ts:10]: done

File: src/base/transport.ts (hunk 2 above)
Comment (iliran11) [reply src/base/transport.ts:22]: done

File: src/base/transport.ts (hunk 3 above)
Comment (iliran11) [reply src/base/transport.ts:26]: Done - add space before each method and labels them as public.

File: src/base/transport.ts (hunk 4 above)
Comment (iliran11) [reply src/base/transport.ts:36]: Done

File: src/base/transport.ts (hunk 5 above)
Comment (iliran11) [reply src/base/transport.ts:84]: Done

File: src/generator/writeClients.ts (hunk 6 above)
Comment (iliran11) [reply src/generator/writeClients.ts:96]: Donne

File: src/generator/writeClients.ts (hunk 7 above)
Comment (iliran11) [reply src/generator/writeClients.ts:100]: Done

**Comment by iliran11** [comment]
Fixed the requests @Silind

**Review by Silind — APPROVED** [review]
Awesome!! :fire:

**Pull request merged**